accepted/00000-http-api.md

```diff
@@ -0,0 +1,214 @@
+
+get("/activationkey/get_details", (req, res) -> {
+  // 1.Unwrap the session user, and the querystring parameter named "key"
+  // 2.Call to the `getDetails` method, passing the unwrapped arguments
+  // 3.Serialize and return the JSON result
+});
+```
+
+### Example 2
+
+Consider the following XMLRPC handler method:
+```java
+public ContentProject updateProject(User user, String label, Map<String, Object> props)
+```
+
+With the proposed annotations, it may be written as:
+```java
+@XmlRpcApiEndpoint
+@HttpApiEndpoint
+public ContentProject updateProject(User user, @RouteParam String label, Map<String, Object> props)
+```
+
+At registration, Spark route wrapper will create the following route definition for this method:
+
+```java
+// POST https://uyuni.server/rhn/manager/api/contentmanagement/update_project/myproject
+
+post("/contentmanagement/update_project/:label", (req, res) -> {
+  // 1.Unwrap the session user, the route parameter named "label",
+  //   and the JSON object named "props" from the request body
+  // 2.Call to the `updateProject` method, passing the unwrapped arguments
+  // 3.Serialize and return the JSON result
+});
+```
+
+
+## 2. Authentication
+
+The web UI performs token-based authentication via a session cookie called `pxt-session-cookie`. Currently, the web UI login page exposes an API endpoint to support this process. This endpoint shall also be used for API authentication.
+
+### Login handshake demo
+
+```bash
+$ curl --include --header "Content-Type: application/json" \
+--data "{'login':'admin','password':'admin'}" \
+https://uyuni.dallas/rhn/manager/api/login
+
+HTTP/1.1 200 200
+...
+Set-Cookie: pxt-session-cookie=209x9fa00e2e947ace87de0743f6cf63b8b0b4bfbfca2310ce6829941fa9db947987; Max-Age=3600; Expires=Tue, 08-Feb-2022 20:05:41 GMT; Path=/; Secure; HttpOnly;HttpOnly;Secure
+...
+
+$ curl --include --cookie "pxt-session-cookie=209x9fa0..." \
+https://uyuni.server/rhn/manager/api/systems/keys
+
+HTTP/1.1 200 200
+...
+<JSON data follows>
+```
+
+
+## 3. I/O
+
+On the HTTP API side, I/O serialization is mostly handled by Spark, and the existing `GSON` library. XMLRPC uses specialized "serializer" classes to write structured output on XML. These serializer classes shall be used exclusively for the XMLRPC output.
+
+
+## 4. Exception Handling
+
+The current implementation of the business logic uses exceptions that are designed specifically for XMLRPC (see: `FaultException` and subclasses). To be able to use the business logic with both interfaces, a new, internal exception structure shall be designed to offer enough flexibility in exception handling. These exceptions are then translated to the specific interface, and processed accordingly. For HTTP endpoints, the internal exceptions can be represented by different HTTP status codes where appropriate.
+
+This is the main point that couples the business logic with a specific interface, and will require the most work to implement.
```

Comment:
The changes in Exceptions is a bit unspecific. I wonder if it is not possible to use the existing exceptions and just enhance them with a HTTP status code. Depending on what layer is used to return it, it takes the fault or the http status.

Reply:
Or could the spark API map them directly to an HTTP status code automaticly? 
In what will a new structure differ from the current one?
Could we add a payload (as json) to the error response HTTP server would send with the data `FaultException` already have?

Reply:
I will export and study the whole exception tree in the API context, then I'll update this section with more specific info.

Reply:
The exception structure is simpler than I thought. We have only one level of inheritance (all from `FaultException`) and there's no XMLRPC-specific attribute in any of the exceptions other than the error number.

> I wonder if it is not possible to use the existing exceptions and just enhance them with a HTTP status code. Depending on what layer is used to return it, it takes the fault or the http status.

Yes, I think this will be possible.

So the question is if we should allow people to arbitrarily assign any HTTP code to any exception they create in the future? Or should we implement some OOP structure to put some ground rules? We probably don't want to end up in an inconsistent mess as we have with the fault codes.

Reply:
Here's a summary of what we have in XMLRPC so far:
https://github.com/uyuni-project/uyuni-rfc/blob/http-api/accepted/attachments/00000-exception-layout.md

Reply:
I think it would make sense to have a-la an enum of status codes we support, that way we can also standardize on the frontend, and then we can provide additional information in a separate field if applicable. I wouldn't allow [the full range of status codes](https://en.wikipedia.org/wiki/List_of_HTTP_status_codes) since it's too large and very subjective. Looking through the XMLRPC exception list, I would say _most_ exceptions can be mapped to one of the following: 

 - 200 OK for all successes, I wouldn't really bother differentiating between 200 and 201 for example etc.
 - 400 Bad Request for all cases where the input values make no sense (if I'm not misunderstanding, then a-la InvalidParameterException etc would fall in this camp?)
 - 401 Unauthorized for missing/incorrect authentication
 - 403 Forbidden for permission issues (a-la NotPermittedByOrgException if I'm not misunderstanding?)
 - 404 Not Found
 - 500 Internal Server Error for other problems, again I don't think it makes sense to differentiate between a-la 500 and 503 etc.

Reply:
Updated the text with more specific info